planner/core/rule_join_reorder_dp.go

```diff
@@ -0,0 +1,194 @@
+		if bits.OnesCount(state) == 1 {
+			continue
+		}
+		// This loop can iterate all its subset.
+		for sub := (state - 1) & state; sub > 0; sub = (sub - 1) & state {
+			remain := state ^ sub
+			if sub > remain {
+				continue
+			}
+			// If this subset is not connected skip it.
+			if bestPlan[sub] == nil || bestPlan[remain] == nil {
+				continue
+			}
+			// Get the edge connecting the two parts.
+			usedEdges := s.nodesAreConnected(sub, remain, oldPos2NewPos, newPos2OldPos, totalEdges)
+			if len(usedEdges) == 0 {
+				continue
+			}
+			join, err := s.newJoinWithEdge(bestPlan[sub], bestPlan[remain], usedEdges)
+			if err != nil {
+				return nil, err
+			}
+			if bestPlan[state] == nil || bestCost[state] > join.statsInfo().Count()+bestCost[remain]+bestCost[sub] {
+				bestPlan[state] = join
+				bestCost[state] = join.statsInfo().Count() + bestCost[remain] + bestCost[sub]
+			}
+		}
+	}
+	return bestPlan[(1<<uint(nodeCnt))-1], nil
+}
+
+func (s *joinReorderDPSolver) nodesAreConnected(leftMask, rightMask uint, oldPos2NewPos, newPos2OldPos []int, totalEdges []joinGroupEdge) []joinGroupEdge {
+	var usedEdges []joinGroupEdge
+	for _, edge := range totalEdges {
+		lIdx := uint(oldPos2NewPos[edge.nodeIDs[0]])
+		rIdx := uint(oldPos2NewPos[edge.nodeIDs[1]])
+		if (leftMask&(1<<lIdx)) > 0 && (rightMask&(1<<rIdx)) > 0 {
+			usedEdges = append(usedEdges, edge)
+		} else if (leftMask&(1<<rIdx)) > 0 && (rightMask&(1<<lIdx)) > 0 {
+			usedEdges = append(usedEdges, edge)
+		}
+	}
+	return usedEdges
+}
+
+func (s *joinReorderDPSolver) newJoinWithEdge(leftPlan, rightPlan LogicalPlan, edges []joinGroupEdge) (LogicalPlan, error) {
+	var eqConds []*expression.ScalarFunction
+	for _, edge := range edges {
+		lCol := edge.edge.GetArgs()[0].(*expression.Column)
+		rCol := edge.edge.GetArgs()[1].(*expression.Column)
+		if leftPlan.Schema().Contains(lCol) {
+			eqConds = append(eqConds, edge.edge)
+		} else {
+			newSf := expression.NewFunctionInternal(s.ctx, ast.EQ, edge.edge.GetType(), rCol, lCol).(*expression.ScalarFunction)
+			eqConds = append(eqConds, newSf)
+		}
+	}
+	join := s.newJoin(leftPlan, rightPlan, eqConds)
+	_, err := join.deriveStats()
+	return join, err
+}
+
+// Make cartesian join as bushy tree.
+func (s *joinReorderDPSolver) makeBushyJoin(cartesianJoinGroup []LogicalPlan) LogicalPlan {
+	for len(cartesianJoinGroup) > 1 {
+		resultJoinGroup := make([]LogicalPlan, 0, len(cartesianJoinGroup))
+		for i := 0; i < len(cartesianJoinGroup); i += 2 {
+			if i+1 == len(cartesianJoinGroup) {
+				resultJoinGroup = append(resultJoinGroup, cartesianJoinGroup[i])
+				break
+			}
+			resultJoinGroup = append(resultJoinGroup, s.newJoin(cartesianJoinGroup[i], cartesianJoinGroup[i+1], nil))
+		}
+		cartesianJoinGroup = resultJoinGroup
+	}
+	return cartesianJoinGroup[0]
+}
+
+func (s *joinReorderDPSolver) newJoinWithCond(lChild, rChild LogicalPlan, eqConds []*expression.ScalarFunction) LogicalPlan {
```

Comment:
Not used?

Reply:
When integrated in planner, there'll be `joinReorderDPSolver.newJoin = newJoinWithCond`.

Reply:
Then why make it a member function of `joinReorderDPSolver`?

Reply:
Emm, got it. Removed